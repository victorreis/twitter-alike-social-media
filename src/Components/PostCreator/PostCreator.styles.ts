import styled from 'styled-components';

export const PostCreatorContainer = styled.div`
  display: flex;
  flex-direction: column;
  padding: 16px;

  border-bottom-width: 1px;
  border-bottom-style: solid;
  border-bottom-color: ${({ theme }) => theme.colors.background.default.normal};
`;

export const PostCreatorHeader = styled.div`
  display: flex;
  column-gap: 1rem;
  justify-content: flex-end;
`;

export const PostCreatorFooter = styled.div`
  padding-top: 16px;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  column-gap: 1rem;
`;
