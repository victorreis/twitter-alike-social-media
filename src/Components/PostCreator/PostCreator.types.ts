import { TestProps } from '../../Config/Tests/Test.types';
import { AvatarClickProps, RequiredAvatarProps } from '../Avatar';

export interface RequiredPostCreatorProps {
  /**
   * Callback function that is called when the post button is clicked.
   */
  onSubmit: (text: string) => void;
}

export interface DefaultPostCreatorProps {}

export interface OptionalPostCreatorProps {}

export type PostCreatorProps = RequiredPostCreatorProps &
  DefaultPostCreatorProps &
  OptionalPostCreatorProps &
  TestProps &
  RequiredAvatarProps &
  AvatarClickProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children' | 'onSubmit'>;

export type PostCreatorStyleProps = Required<DefaultPostCreatorProps>;
