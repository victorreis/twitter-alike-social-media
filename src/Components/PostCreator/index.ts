export * from './PostCreator.types';
export * from './PostCreator';
