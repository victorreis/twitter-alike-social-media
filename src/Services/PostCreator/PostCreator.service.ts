import dayjs from 'dayjs';
import { nanoid } from 'nanoid';

import {
  LOCAL_STORAGE,
  SHORT_DATE_FORMAT,
} from '../../Config/Constants.config';
import {
  REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY,
  USER_NOT_FOUND,
} from '../../Config/ErrorMessages.config';
import { PostType } from '../../Models/Post.types';
import { QuotePostType } from '../../Models/QuotePost.types';
import { RepostType } from '../../Models/Repost.types';
import { postRetrieverService } from '../PostRetriever';
import { userRetrieverService } from '../UserRetriever';
import { PostCreatorService } from './PostCreator.service.type';

const reachedTheMaximumNumberOfPostsPerDay = (userId: string) => {
  const postsByUser = postRetrieverService.getAllCreatedByUser(userId);
  const todayPosts = postsByUser.filter(
    (post) =>
      dayjs(post.createdAt).format(SHORT_DATE_FORMAT) ===
      dayjs(new Date()).format(SHORT_DATE_FORMAT)
  );

  return todayPosts.length >= 5;
};

const createPost = (text: string, createdById: string): PostType => {
  const posts = postRetrieverService.getAll();
  const userCreator = userRetrieverService.getById(createdById);

  if (!userCreator) {
    throw Error(USER_NOT_FOUND(createdById));
  }
  if (reachedTheMaximumNumberOfPostsPerDay(createdById)) {
    console.error(REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY);
    throw Error(REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY);
  }

  const newPost: PostType = {
    type: 'Post',
    id: nanoid(),
    text,
    createdAt: new Date(),
    createdBy: userCreator,
    reposts: 0,
    quotPosts: 0,
  };

  posts.push(newPost);
  localStorage.setItem(LOCAL_STORAGE.POSTS, JSON.stringify(posts));

  return newPost;
};

const createRepost = (
  originalPostId: string,
  createdById: string
): RepostType => {
  const posts = postRetrieverService.getAll();
  const originalPost = postRetrieverService.getById(originalPostId) as PostType;
  const userCreator = userRetrieverService.getById(createdById);

  if (!originalPost || !userCreator) {
    throw Error(USER_NOT_FOUND(createdById));
  }
  if (reachedTheMaximumNumberOfPostsPerDay(createdById)) {
    console.error(REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY);
    throw Error(REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY);
  }

  const newRepost: RepostType = {
    type: 'Repost',
    originalPost,
    id: nanoid(),
    createdAt: new Date(),
    createdBy: userCreator,
  };

  posts.push(newRepost);
  localStorage.setItem(LOCAL_STORAGE.POSTS, JSON.stringify(posts));

  return newRepost;
};

const createQuotePost = (
  originalPostId: string,
  text: string,
  createdById: string
): QuotePostType => {
  const posts = postRetrieverService.getAll();
  const originalPost = postRetrieverService.getById(originalPostId) as PostType;
  const userCreator = userRetrieverService.getById(createdById);

  if (!userCreator) {
    throw Error(USER_NOT_FOUND(createdById));
  }
  if (reachedTheMaximumNumberOfPostsPerDay(createdById)) {
    console.error(REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY);
    throw Error(REACHED_THE_MAX_NUMBER_OF_POSTS_PER_DAY);
  }

  const newPost: QuotePostType = {
    type: 'QuotePost',
    originalPost,
    id: nanoid(),
    text,
    createdAt: new Date(),
    createdBy: userCreator,
    reposts: 0,
    quotPosts: 0,
  };

  posts.push(newPost);
  localStorage.setItem(LOCAL_STORAGE.POSTS, JSON.stringify(posts));

  return newPost;
};

export const postCreatorService: PostCreatorService = {
  createPost,
  createRepost,
  createQuotePost,
};
