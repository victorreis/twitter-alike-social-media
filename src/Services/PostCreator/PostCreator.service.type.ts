import { PostType } from '../../Models/Post.types';
import { QuotePostType } from '../../Models/QuotePost.types';
import { RepostType } from '../../Models/Repost.types';

export interface PostCreatorService {
  createPost: (text: string, createdById: string) => PostType;

  createRepost: (originalPostId: string, createdById: string) => RepostType;

  createQuotePost: (
    originalPostId: string,
    text: string,
    createdById: string
  ) => QuotePostType;
}
