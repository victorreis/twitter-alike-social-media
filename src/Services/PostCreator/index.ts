export * from './PostCreator.service';
export * from './PostCreator.service.type';
