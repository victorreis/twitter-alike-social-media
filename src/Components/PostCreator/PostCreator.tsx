import {
  forwardRef,
  ForwardRefRenderFunction,
  useState,
  useCallback,
} from 'react';

import { POST_CREATOR_MAX_LENGTH } from '../../Config/Constants.config';
import { TestProps } from '../../Config/Tests/Test.types';
import { useShowUserPage } from '../../Hooks/ShowUserPage';
import { Avatar } from '../Avatar';
import { Button } from '../Button';
import { Textarea } from '../Textarea';
import { Typography } from '../Typography';
import {
  PostCreatorContainer,
  PostCreatorFooter,
  PostCreatorHeader,
} from './PostCreator.styles';
import { PostCreatorProps, DefaultPostCreatorProps } from './PostCreator.types';

export const postCreatorDefaults: Required<DefaultPostCreatorProps> &
  Required<TestProps> = {
  testID: 'PostCreator',
};

const PostCreatorComponent: ForwardRefRenderFunction<
  HTMLTextAreaElement,
  PostCreatorProps
> = (props, ref) => {
  const {
    testID = postCreatorDefaults.testID,
    thumbnailUrl,
    name,
    nickname,
    onSubmit,
    style,
    ...others
  } = props;

  const [value, setValue] = useState('');
  const [characterCounter, setCharacterCounter] = useState(0);

  const { handleShowUserPage, isAvatarClickable } = useShowUserPage({
    nickname,
  });

  const handleChange = (newValue: string) => {
    setValue(newValue);
    setCharacterCounter(newValue.length);
  };

  const handleSubmit = useCallback(() => {
    if (value) {
      setValue(() => '');
      onSubmit(value);
      setCharacterCounter(() => 0);
    }
  }, [onSubmit, value]);

  return (
    <PostCreatorContainer data-testid={testID} style={style} {...others}>
      <PostCreatorHeader>
        <Avatar
          clickable={isAvatarClickable}
          name={name}
          onClick={handleShowUserPage}
          thumbnailUrl={thumbnailUrl}
        />
        <Textarea
          ref={ref}
          maxLength={POST_CREATOR_MAX_LENGTH}
          onChange={handleChange}
          placeholder="What is happening?"
          value={value}
        />
      </PostCreatorHeader>

      <PostCreatorFooter>
        <Typography>
          {characterCounter} / {POST_CREATOR_MAX_LENGTH}{' '}
        </Typography>
        <Button onClick={handleSubmit}>POST</Button>
      </PostCreatorFooter>
    </PostCreatorContainer>
  );
};

export const PostCreator = forwardRef(PostCreatorComponent);
