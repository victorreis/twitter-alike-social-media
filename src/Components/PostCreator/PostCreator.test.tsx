import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { PostCreator, postCreatorDefaults } from './PostCreator';
import { PostCreatorProps } from './PostCreator.types';

describe('postCreator component tests', () => {
  const onSubmit = jest.fn();
  const thumbnailUrl =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAcCAYAAAATFf3WAAAAAXNSR0IArs4c6QAAArJJREFUWEftlVtIk2EYx3/bPKERZOkItUSIskLojKDRwRrDIlMo0wkzwwmpQzS7KPBQ1E3zstZuUjONsIvAxNKMJPDCykM1yQ5WHppmIeWhMl1839xyUG4xJS/28V297/O+///3ew6fxGw2m1nAj8Rt0MXsuAm6CBA3QTdBVwm4et5BDVpnuMROx4z9bJfgaF84bh9juVC450/rv+UcNsn4+BdytLn8nLI3lZauoa2lmcet7TaRxCQVu3dEYTAYuF1bh6+vH5laLX1vX1J/74FNVbgpOCSUgoJTSF0xKJD6Pj5CYeEZRkdHuHRRzx5lLOvC13BElcTpk7l0ve7nYNw+UfxAfAKDPW84lKzmRH4+TE6wyD+AiNVhND1sprGhnve9A6jVKpYHhZCTk4UU6axVMCtBCzOz+A5/MiGXB6MvvUZqymEkTKJU7AXPJdTWVIschN/6+eICis7paH3aQfiqMIu4xJLG7Ix06pseYTQ+EeabSF4yvfc3lw5TbD04PGQiUB7EZdFgIjCFUhFDa8cLtm3djFTmRenVMvq6u4iO3sm3HxMkq1IoPluEPGCZaCYrI52GphaMxjYH3P6hBh0ZNH0cIy8vG6lURnxCPF4eMnp7eyjR6dDrDWyJ3M79xjvIJJA5bbDT2Oag8ubQIJ7+YorFYjBPMT42iq+fn1gW2uMayipvMTDUh4+Hx/8x2P7sFdFRkaLBpBQ1NTcrGfz8ldCVK7heVcUuRSyVFVfEfWuK54Gg0M1jXNCVsD8ugYj1a0XB6htVPO/ssuUjRqEkcOliyssr+GAaYMPGTaQdTcXH2xthItytq6X7XT8azTGhPZya4U43iWWoii05Y56Z7YSsA1zsaFvkTCOWVSFuHgw69cFzHvQPBOdc26kL3QadwjRLkJugm6CrBFw9v+Br8BdSpWm8K4SiiQAAAABJRU5ErkJggg==';
  const name = 'Victor Reis';
  const nickname = 'victor.reis';

  const requiredProps: PostCreatorProps = {
    onSubmit,
    thumbnailUrl,
    name,
    nickname,
  };

  const setup = (props?: PostCreatorProps) => {
    const renderRTR = () =>
      renderRTRCreator<PostCreatorProps>(PostCreator, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<PostCreatorProps>(PostCreator, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(postCreatorDefaults.testID);

      expect(testInstance).toBeTruthy();
    });
  });

  describe('style tests', () => {});

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
