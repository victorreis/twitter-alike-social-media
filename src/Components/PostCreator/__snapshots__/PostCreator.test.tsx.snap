// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postCreator component tests snapshot tests should render correctly 1`] = `
<div
  className="PostCreatorContainer-twitter-alike-social-media__sc-r64v60-0 glllXS"
  data-testid="PostCreator"
  onChange={[MockFunction]}
>
  <div
    className="PostCreatorHeader-twitter-alike-social-media__sc-r64v60-1 beCQWi"
  >
    <div
      className="AvatarContainer-twitter-alike-social-media__sc-ne4tqs-0 gejhQE"
      data-testid="Avatar"
      size="MD"
    >
      <img
        alt="Victor Reis's avatar photo."
        className="AvatarImage-twitter-alike-social-media__sc-ne4tqs-1 dnbueO"
        data-testid="Avatar_image"
        onError={[Function]}
        size="MD"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAcCAYAAAATFf3WAAAAAXNSR0IArs4c6QAAArJJREFUWEftlVtIk2EYx3/bPKERZOkItUSIskLojKDRwRrDIlMo0wkzwwmpQzS7KPBQ1E3zstZuUjONsIvAxNKMJPDCykM1yQ5WHppmIeWhMl1839xyUG4xJS/28V297/O+///3ew6fxGw2m1nAj8Rt0MXsuAm6CBA3QTdBVwm4et5BDVpnuMROx4z9bJfgaF84bh9juVC450/rv+UcNsn4+BdytLn8nLI3lZauoa2lmcet7TaRxCQVu3dEYTAYuF1bh6+vH5laLX1vX1J/74FNVbgpOCSUgoJTSF0xKJD6Pj5CYeEZRkdHuHRRzx5lLOvC13BElcTpk7l0ve7nYNw+UfxAfAKDPW84lKzmRH4+TE6wyD+AiNVhND1sprGhnve9A6jVKpYHhZCTk4UU6axVMCtBCzOz+A5/MiGXB6MvvUZqymEkTKJU7AXPJdTWVIschN/6+eICis7paH3aQfiqMIu4xJLG7Ix06pseYTQ+EeabSF4yvfc3lw5TbD04PGQiUB7EZdFgIjCFUhFDa8cLtm3djFTmRenVMvq6u4iO3sm3HxMkq1IoPluEPGCZaCYrI52GphaMxjYH3P6hBh0ZNH0cIy8vG6lURnxCPF4eMnp7eyjR6dDrDWyJ3M79xjvIJJA5bbDT2Oag8ubQIJ7+YorFYjBPMT42iq+fn1gW2uMayipvMTDUh4+Hx/8x2P7sFdFRkaLBpBQ1NTcrGfz8ldCVK7heVcUuRSyVFVfEfWuK54Gg0M1jXNCVsD8ugYj1a0XB6htVPO/ssuUjRqEkcOliyssr+GAaYMPGTaQdTcXH2xthItytq6X7XT8azTGhPZya4U43iWWoii05Y56Z7YSsA1zsaFvkTCOWVSFuHgw69cFzHvQPBOdc26kL3QadwjRLkJugm6CrBFw9v+Br8BdSpWm8K4SiiQAAAABJRU5ErkJggg=="
        title="Victor Reis's avatar photo."
      />
    </div>
    <textarea
      className="StyledTextarea-twitter-alike-social-media__sc-12f32i7-0 ioqcrC"
      data-testid="Textarea"
      maxLength={777}
      onChange={[Function]}
      placeholder="What is happening?"
      rows={4}
      value=""
    />
  </div>
  <div
    className="PostCreatorFooter-twitter-alike-social-media__sc-r64v60-2 gtgaaP"
  >
    <span
      className="TypographyContainer-twitter-alike-social-media__sc-1rk4wyq-0 bsZKYT"
      data-testid="Typography"
    >
      0
       / 
      777
       
    </span>
    <button
      className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 fBnyBV"
      data-testid="Button"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      size="LG"
      type="button"
    >
      POST
    </button>
  </div>
</div>
`;
